File: 001_multiples_of_three_and_five/multiples_of_three_and_five.rb
# Finds multiples of three or five below a given number, 
#   and finds the sum of the multiples
def sum_of_multiples(below_num)
  sum = 0
  for i in 0...below_num do
    sum += i if i % 3 == 0 || i % 5 == 0
  end
  sum
end

# Tracks time
start = Time.now
puts sum_of_multiples(1000)
puts Time.now - start
# Answer: 233168
# Time Elapsed: 0.001

File: 001_multiples_of_three_and_five/multiples_of_three_and_five.ex
# Finds multiples of three or five below a given number, 
#   and finds the sum of the multiples
defmodule MultiplesThreeFive do
  def sum(below_num) do
    range = 1..(below_num - 1)
    IO.puts Enum.filter(range, fn(x) -> rem(x, 3) == 0 || rem(x, 5) == 0 end)
      |> Enum.reduce(fn(sum, num) -> sum + num end)
  end
end

# Used to track time
defmodule Benchmark do
  def measure(function) do
    function
    |> :timer.tc
    |> elem(0)
    |> Kernel./(1_000_000)
  end
end

IO.puts Benchmark.measure(fn -> MultiplesThreeFive.sum(1000) end)